### PR TITLE
Add action closures for all UIControl subclasses

### PR DESCRIPTION
**Pull request by apcorc**
screen shot 2016-06-23 at 9 20 44 am

### CODE REVIEW

File: Sources/UIControl+Actionable.swift [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+protocol Actionable: NSObjectProtocol {
+    
+    init(actionClosure: Self -> Void)
```
Comment (jlandon) [inline Sources/UIControl+Actionable.swift]: [SE-0066](https://github.com/apple/swift-evolution/blob/master/proposals/0066-standardize-function-type-syntax.md) requires parentheses around parameter lists, so this (and other cases throughout the PR) should be updated
```swift
// example
init(actionClosure: (Self) -> Void)
```

Comment (apcorc) [inline Sources/UIControl+Actionable.swift]: Ah, that's my preferred style too, I was just keeping with what was in the old `UIButton` version. I'll update.

File: Sources/UIControl+Actionable.swift [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+    }
+    
+    /**
+     Initialize a UIControl, using the given closure as the .TouchUpInside target/action event.
+     
+     - parameter action: The closure to execute upon control press.
+     
+     - returns: An initialized UIControl.
+     */
+    init(actionClosure: (Self) -> Void) {
```
Comment (jlandon) [inline Sources/UIControl+Actionable.swift]: Both initializers and the `addTarget(controlEvents:actionClosure:)` method should be `public`

File: Sources/UIControl+Actionable.swift (hunk 1 above)
Comment (hsoi) [inline Sources/UIControl+Actionable.swift]: Yeah... if @jlandon is already trying to update even style stuff in the `swift3` branch, that's cool. But if you want to have new code start to conform to the new styles/conventions, I see no reason not to do so.

File: Sources/UIControl+Actionable.swift [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+private struct AssociatedKeys {
+    static var ActionName = "action"
+}
+
+@objc protocol ActionSelectable {
+    func handleAction()
+}
+
+extension UIControl: ActionSelectable, Actionable {
+    func handleAction() {
```
Comment (hsoi) [inline Sources/UIControl+Actionable.swift]: Will this need to be `public`?

File: Sources/UIControl+Actionable.swift [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+        guard let action = self.action else {
+            fatalError("\n" +
```
Comment (hsoi) [inline Sources/UIControl+Actionable.swift]: I like this. Very nice way to document.

**Comment by jlandon** [comment]
Just gonna throw this out there, but the code could be greatly simplified if we didn't return the control in the closure. The reason multiple protocols and generics are necessary is simply because we're returning the object. If we were to remove this, the code becomes much simpler, and we don't really lose any functionality. If someone were to want the control when the closure is called, they can easily do so like this:
```swift
button.addTarget(controlEvents: .TouchUpInside) {
   button.enabled = false
   // do other stuff with `button`
}
```
Thoughts?

**Comment by apcorc** [comment]
@jlandon I thought about that, and you're right it's very much simplified, I just like the ability to golf the control at `$0` to get things like `value` which is especially important in controls other than `UIButton` which this PR is designed to support. Also if we remove the control, we'd need to keep the old implementation for UIButton and deprecate it due to code in use now which expects the control to be returned.

**Comment by apcorc** [comment]
After looking into it more we've come down to an option: We can do as @jlandon suggested and not return the control in the closure and support all UIControl subclasses _INCLUDING_ custom UIControl subclasses that you define in your own projects outside what Alexandria knows about. Or we can continue passing the control back as a parameter in the closure but only support UIKit and Alexandria custom controls that are defined ahead of time. My idea of just adding the extension to support a custom control didn't work as `self.action` is undefined at that scope and can't be made public because of the use of `Self`. Or we can do both and only pass back the control if we know about it maybe? I'm interested in ya'll thoughts

**Comment by hsoi** [comment]
https://www.youtube.com/watch?v=CxK_nA2iVXw

**Comment by hsoi** [comment]
My Neutralness aside....

Given that Alexandria is public to the world, I think it makes better sense to serve that audience. So being able to support custom subclasses outside of Alexandria's knowledge will probably serve best in the long run.

**Comment by jlandon** [comment]
Yeah, I feel like the behavior should be the same for all UIControl subclasses, so if there's not a way to extend the behavior to built-in UIKit controls, I think we should go with the closure that doesn't return the control

**Comment by jlandon** [comment]
Now that we've done away with the generic, I'd say we just extend UIControl directly and remove the `Actionable` protocol

**Comment by hsoi** [comment]
I'll agree there, because what's in the protocol is inherently `UIControl`-specific, so I'm not sure what's gained.

If there is something @apcorc please enlighten.

**Comment by apcorc** [comment]
Ah yea I do remember talking about that before. Should be fine, just forgot about it.

**Comment by jlandon** [comment]
👍

**Pull request merged**